### ISSUE TITLE
Support connecting to networks that have a captive portal

### ISSUE THREAD

**Issue by pcarranzav** [resin-io/resin-wifi-connect#51]
Probably a tricky one, but we could either support:
* Using WISPr to authenticate with portals that support it, or
* Having a simple web crawler that enters the credentials into the correct fields. Maybe the field names can be specified (with env vars?) and instead of a crawler we just do a simple HTTP request to authenticate.

**Comment by ab77** [comment]
Yes, this is really a show stopper for places such as hotels and public Wi-Fi hotspots, since there is no way to authenticate against an open network after it has been joined (no passphrase).

**Comment by majorz** [comment]
I will raise this for this discussion with our team to see what is their opinion for possible implementation approaches. My initial thinking is that we need to open WiFi Connect with some form of plugin support and this suits that category well. That said, it will take a while to get there, but I will post some initial feedback from the team once we discuss it next week.

**Comment by alexandrosm** [comment]
if we can spot that the network we've connected to has a captive portal
itself, theoretically we could somehow forward that HTML to the user and
replay their actions on their behalf or something. But this isn't trivial.

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by majorz** [comment]
@alexandrosm NetworkManager has built-in captive portal detection. It has to be enabled in the global NetworkManager configuration though, since it requires additional request to a specified address (like our ping endpoint maybe). This also relates to reconfix handling our global NetworkManager configuration.

Captive portal detection will be easy to enable in WiFi Connect, since we already check for connectivity after a connection is established - we just need to check additionally for the `NM_CONNECTIVITY_PORTAL` flag.

However dealing with networks behind a captive portal does not have much to do with what WiFi Connect does now - setting up a new NetworkManager connection profile. Once a wireless connection to a network behind a captive portal is setup, the user will still have to re-authenticate with the captive portal at later stages, e.g. if the captive portal session has expired or the device is rebooted.

That means that there should be a long living application that re-authenticates periodically with the captive portal.

If captive portal type of connectivity is detected WiFi Connect can spawn a user provided application - git client hooks come to my mind here as a possible approach. That application could then communicate with the captive portal dealing with cookies, filling form data, checking for expired sessions, etc. - all of that being non-trivial. I do not feel WiFi Connect itself should be expanded with that, but we can eventually provide hooks for user applications since we have other more practical use cases that will benefit from this.

**Comment by alexandrosm** [comment]
I suppose if we're being ambitious, wifi connect could store the movements
the user makes on the captive portal page as part of the authentication
information, and replay those movements every time it needs to reconnect.
That would be the ultimate solution to the problem, but it's not an easy
one of course.

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by majorz** [comment]
That has other limitations as well. Before we connect to the desired network, our own access point with a captive portal is closed and the user is disconnected from our application. That is to say we cannot display the captive portal of the desired network through the captive portal of WiFi Connect. Something like that can only happen if there are two wireless devices present on the system, and we run our access point on one of them and the other wireless device is connected to the desired network. With a single wireless device that is not possible, even if the device supports it, because of the limitations of NetworkManager (works only with hostapd). I like the idea of capturing and replaying though :D

**Comment by alexandrosm** [comment]
On some devices it's possible to have hotspot and be connected at the same
time. This would be needed to handle captive portals.

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by majorz** [comment]
Not possible under NetworkManager unfortunately. I investigated that recently in detail, since we already had such request here and a user on the forums was having the same question.

NetworkManager cannot create an AP with the device being in client/station + access point mode (STA+AP). It has a very fixed flow for how to handle newly added interfaces. To enter STA+AP mode a virtual interface of type `__ap` has to be added like this: `iw dev wlan0 interface add ap0 type __ap`. After the interface is added NetworkManager forces the virtual interface to enter station/client mode, which does not work, since the interface is created with AP type.

This is the reason in all guides NetworkManager is set to not manage the particular interface and hostapd is used instead. I checked with latest version as well, but still no support available.

**Comment by alexandrosm** [comment]
You should make sure to ping @petrosagg on this, he may have ideas.

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by majorz** [comment]
Right, let's do that. I think it will be best to include it in the arch call when we discuss the two WiFi Connect issues I raised this week (about this and another ticket).

**Comment by majorz** [comment]
Just a quick update on this after the discussion this week: I will spend more time in investigating how NetworkManager can support STA+AP mode, so that we have a more clear view on what is possible to accomplish with a single adapter.

**Comment by ab77** [comment]
@majorz FWIIW, this works on a NUC, putting the built-in WLAN interface into both a station and AP mode simultaneously (STA+AP), using  a pair of virtual interfaces ([guide](https://superuser.com/questions/649220/a-wifi-ap-with-a-single-nic)).

I've tested below on an `resin/%%RESIN_MACHINE_NAME%%-alpine-python:slim`.

```
export DBUS_SYSTEM_BUS_ADDRESS=unix:path=/host/run/dbus/system_bus_socket
service networkmanager stop
iw dev wlp2s0 del
iw phy phy0 interface add new0 type station
service networkmanager start
iw phy phy0 interface add new1 type __ap

# cat /etc/dnsmasq.d/dnsmasq.conf
interface=new1
bind-dynamic
log-facility=-
no-resolv
server=8.8.8.8
dhcp-range=172.24.1.50,172.24.1.150,12h

# cat /etc/hostapd.conf
interface=new1
ssid={{ssid}}
auth_algs=1
wpa=2
wpa_key_mgmt=WPA-PSK
wpa_passphrase={{passphrase}}
wpa_pairwise=CCMP TKIP
rsn_pairwise=CCMP
ieee80211n=1
ieee80211ac=1
hw_mode=g
channel=1

# hostapd /etc/hostapd.conf &
# ifconfig new1 172.24.255.254
# dnsmasq &

# nmcli device wifi connect {{ssid}} password {{passphrase}} ifname new0

# nmcli d
DEVICE     TYPE      STATE        CONNECTION
resin-dns  bridge    connected    resin-dns
new0       wifi      connected    {{ssid}}
enp3s0     ethernet  unavailable  --
new1       wifi      unavailable  --
...

# ip route
default via 192.168.x.y dev new0 proto dhcp metric 600
...
```

At this point, both the AP is setup and the devices is connected to the internet using it's Wi-Fi connection (Ethernet unplugged). If forwarding and NAT are enabled, clients connected to AP `new1` interface are forwarded out the `new0` interface to the network.

**Comment by majorz** [comment]
@ab77 Thanks for posting this - very nice! I am planning to continue the related work soon. First thing will be to dive more into the details of what exactly is needed for NetworkManager to support this without hostapd. This will need digging into the NetworkManager's source code more and tracing the calls. I will post more information here once I start working on this.

**Comment by ab77** [comment]
Awesome, it would definitely be a bonus if all of this can be done with NM alone.

**Comment by jhabdas** [comment]
Besides the fact I just tried this on a RPi3 and never saw a hotspot created I figured I'd save you all a little time and mention you can bypass captive portals by spoofing the device MAC address using macOS. Instructions here: https://hackcabin.com/post/spoofing-iot-device-mac-addresses/

**Comment by drjasonharrison-vp-eio** [comment]
@majorz any progress on this. We're finding that public wifi networks with "terms and conditions" which I believe are captive portals are keeping us from getting internet access.

How would we know if the wifi network has a captive portal? Assume we can connect using a laptop, which tools would we use?

**Comment by ab77** [comment]
Make a HTTPS cURL request to your site and check for response code you are expecting (i.e. 200). If you get a 302 or something else, assume captive portal.

**Comment by majorz** [comment]
The biggest limitation for this type of feature is that NetworkManager does not support being connected to a network and having an access point at the same time. If we are able to resolve this in the future, we can think of possible ways forward, but up until then I would not recommend usage of WiFi Connect on setups where support for public networks with captive portals is a must.

**Comment by drjasonharrison-vp-eio** [comment]
@majorz thank you for the advice. Do you have an alternative suggestion for a wifi-connect like package that may support captive portals?

We will also look into MAC spoofing.

**Comment by majorz** [comment]
@drjasonharrison-vp-eio I really doubt such application exists, but if you find anything please let me know as I am interested in this.

**Comment by silvia-eio** [comment]
A way we found to at least validate if it's a captive portal, and allow a new connection is through ``` wget "http://clients3.google.com/generate_204?" -O /dev/null 2>&1 | grep "204 No Content" > /dev/null ```

**Comment by rhwood** [comment]
[NetworkManager issue 58](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/issues/58) covers the need to support STA+AP in NetworkManager.

**Comment by andrewjaykeller** [comment]
Any update on this? Device unable to join Fly-FI on JetBlue and now I’m locked out.

**Comment by evanlesmez** [comment]
What is the status on this? [The Network Manager issue](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/58) was auto-closed but acknowledges a ["hack" solution in the archwiki](https://wiki.archlinux.org/title/Software_access_point).

**Comment by janvda** [comment]
Is there any update ?
I cannot consult above jel.ly.fish links as you need to logon.

E.g. I would like to use it to connect to a [kpn hotspot](https://www.kpn.com/internet/wifi-hotspots/premium.htm) using an access code we have received.

<img width="719" alt="image" src="https://github.com/balena-os/wifi-connect/assets/17238398/f64f4940-295c-4421-b699-ae7d92da02c7">

**Comment by drjasonharrison** [comment]
There is no standard for captive hot spot portals. Many IoT devices cannot connect to these portals because of the lack of standardization. Hotels require a name, room number, and daily authentication. Some cafes require an "I agree" button or checkbox. Some require an email address. 

The options include:

- use something like a travel router to provide a connection to the portal and a network signal for a phone or other device with a web browser to complete the connection process. The MAC used by the router is what the captive portal sees.
- use MAC spoofing of the IoT device MAC to connect a phone or computer to the captive portal then power on  the IoT device.
- run a full browser stack (https, DNS, JavaScript...) on the IoT device and using its human user interface complete the connection process.
- using a WiFi and Bluetooth (or other physical network channel) have the IoT device behave like a router that sends all of the portal connection interaction to a phone or computer so the connection process can be completed.
- whitelist the IoT device on the portal network

**Comment by janvda** [comment]
@drjasonharrison  very interesting.

Regarding 2nd option:

> use MAC spoofing of the IoT device MAC to connect a phone or computer to the captive portal then power on the IoT device.

So following steps should work I hope:

1.  Connect with phone or a tablet to wifi network + entering authentication key (or other information) in captive hot spot portal.
2. Once properly connected - switch off the wifi from your phone/tablet
3. Assure that your IoT device is using same MAC address as your phone/tablet (MAC spoofing)
    * this is possible on raspberry pi 3
4. Start your IoT device which will setup a hotspot
5. Connect to the hotspot of the IoT device and select the proper wifi network
     1. I am not sure if you must enter a wifi password.  Anyway the wifi network is not using a wifi password.
